From a79ba2ce5be0eeb862d6802974b71e08c79a1b1a Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Mon, 3 Aug 2026 11:53:23 -0400
Subject: [PATCH 254/268] win32u: Introduce pEnableIMEContext driver call.

Based on a patch by Ziia Shi.
---
 dlls/win32u/driver.c      | 11 +++++++++++
 dlls/win32u/imm.c         |  9 ++++++++-
 include/wine/gdi_driver.h |  1 +
 3 files changed, 20 insertions(+), 1 deletion(-)

diff --git a/dlls/win32u/driver.c b/dlls/win32u/driver.c
index 3c1286b1fdd..793de3fd27e 100644
--- a/dlls/win32u/driver.c
+++ b/dlls/win32u/driver.c
@@ -683,6 +683,10 @@ static BOOL nulldrv_SetIMECompositionRect( HWND hwnd, RECT rect )
     return FALSE;
 }
 
+static void nulldrv_EnableIMEContext( HWND hwnd, BOOL enabled )
+{
+}
+
 static LRESULT nulldrv_DesktopWindowProc( HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam )
 {
     return default_window_proc( hwnd, msg, wparam, lparam, FALSE );
@@ -1116,6 +1120,11 @@ static BOOL loaderdrv_SetIMECompositionRect( HWND hwnd, RECT rect )
     return load_driver()->pSetIMECompositionRect( hwnd, rect );
 }
 
+static void loaderdrv_EnableIMEContext( HWND hwnd, BOOL enabled )
+{
+    return load_driver()->pEnableIMEContext( hwnd, enabled );
+}
+
 static LONG loaderdrv_ChangeDisplaySettings( LPDEVMODEW displays, LPCWSTR primary_name, HWND hwnd,
                                              DWORD flags, LPVOID lparam )
 {
@@ -1266,6 +1275,7 @@ static const struct user_driver_funcs lazy_load_driver =
     loaderdrv_ImeProcessKey,
     loaderdrv_NotifyIMEStatus,
     loaderdrv_SetIMECompositionRect,
+    loaderdrv_EnableIMEContext,
     /* cursor/icon functions */
     nulldrv_DestroyCursorIcon,
     loaderdrv_SetCursor,
@@ -1371,6 +1381,7 @@ void __wine_set_user_driver( const struct user_driver_funcs *funcs, UINT version
     SET_USER_FUNC(ImeProcessKey);
     SET_USER_FUNC(NotifyIMEStatus);
     SET_USER_FUNC(SetIMECompositionRect);
+    SET_USER_FUNC(EnableIMEContext);
     SET_USER_FUNC(DestroyCursorIcon);
     SET_USER_FUNC(SetCursor);
     SET_USER_FUNC(GetCursorPos);
diff --git a/dlls/win32u/imm.c b/dlls/win32u/imm.c
index 3337f66e291..0008a19839d 100644
--- a/dlls/win32u/imm.c
+++ b/dlls/win32u/imm.c
@@ -186,6 +186,7 @@ UINT WINAPI NtUserAssociateInputContext( HWND hwnd, HIMC ctx, ULONG flags )
 {
     WND *win;
     UINT ret = AICR_OK;
+    HIMC old;
 
     TRACE( "%p %p %x\n", hwnd, ctx, flags );
 
@@ -214,12 +215,14 @@ UINT WINAPI NtUserAssociateInputContext( HWND hwnd, HIMC ctx, ULONG flags )
     if (ctx && !is_current_thread_window( hwnd )) return AICR_FAILED;
     if (!(win = get_win_ptr( hwnd )) || win == WND_OTHER_PROCESS || win == WND_DESKTOP) return AICR_FAILED;
 
-    if (flags != IACE_IGNORENOCONTEXT || win->imc)
+    if ((old = win->imc) || flags != IACE_IGNORENOCONTEXT)
     {
         if (win->imc != ctx && get_focus() == hwnd) ret = AICR_FOCUS_CHANGED;
         win->imc = ctx;
     }
 
+    if (!old ^ !win->imc) user_driver->pEnableIMEContext( hwnd, !!win->imc );
+
     release_win_ptr( win );
     return ret;
 }
@@ -313,6 +316,8 @@ BOOL register_imm_window( HWND hwnd )
                                                           0, 0, 1, 1, 0, 0, 0, 0, 0, 0, NULL, FALSE );
     }
 
+    user_driver->pEnableIMEContext( hwnd, TRUE );
+
     return TRUE;
 }
 
@@ -336,6 +341,8 @@ void unregister_imm_window( HWND hwnd )
         HWND destroy_hwnd = detach_default_window( thread_data );
         if (destroy_hwnd) NtUserDestroyWindow( destroy_hwnd );
     }
+
+    user_driver->pEnableIMEContext( hwnd, FALSE );
 }
 
 /***********************************************************************
diff --git a/include/wine/gdi_driver.h b/include/wine/gdi_driver.h
index e41e01e484a..d52f6ec3734 100644
--- a/include/wine/gdi_driver.h
+++ b/include/wine/gdi_driver.h
@@ -383,6 +383,7 @@ struct user_driver_funcs
     UINT    (*pImeProcessKey)(HIMC,UINT,UINT,const BYTE*);
     void    (*pNotifyIMEStatus)(HWND,UINT);
     BOOL    (*pSetIMECompositionRect)(HWND,RECT);
+    void    (*pEnableIMEContext)(HWND, BOOL);
     /* cursor/icon functions */
     void    (*pDestroyCursorIcon)(HCURSOR);
     void    (*pSetCursor)(HWND,HCURSOR);
-- 
2.55.0

